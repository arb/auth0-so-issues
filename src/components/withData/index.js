import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Spinner from '../Spinner';

const Container = ({ children, name }) => (
  <div>
    <h1>{name}</h1>
    {children}
  </div>
);

Container.propTypes = {
  children: PropTypes.element.isRequired,
  name: PropTypes.string.isRequired,
};

const withData = (WrappedComponent, name) => class extends Component {
  static propTypes = {
    shouldFetch: PropTypes.bool.isRequired,
    auth: PropTypes.object.isRequired,
    fetchQuestions: PropTypes.func.isRequired,
    fetch: PropTypes.shape({
      loading: PropTypes.bool.isRequired,
      error: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    }).isRequired,
  };
  componentDidMount() {
    const { shouldFetch, fetchQuestions, auth } = this.props;
    if (shouldFetch) {
      fetchQuestions(auth);
    }
  }
  render() {
    const { loading, error } = this.props.fetch;
    if (loading) {
      return (
        <Container name={name}>
          <Spinner>
            <span>Loading data... Please be patient as there is a lot of data to load!</span>
          </Spinner>
        </Container>
      );
    }

    if (error) {
      return (
        <Container name={name}>
          <span>{error}</span>
        </Container>
      );
    }

    return (
      <Container name={name}>
        <WrappedComponent {...this.props} />
      </Container>
    );
  }
};

export default withData;
