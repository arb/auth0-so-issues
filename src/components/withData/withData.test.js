/* eslint-env jest */
import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import withData from './index';

const Comp = withData(() => (<div>hello world</div>), 'test');

describe('<withData />', () => {
  it('shows a spinner if loading is true', () => {
    const wrapper = shallow(<Comp fetch={{
      loading: true,
      error: null,
    }}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('shows an error if present', () => {
    const wrapper = shallow(<Comp fetch={{
      loading: false,
      error: 'There was an error getting data',
    }}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders the component only if there are no errors and loading is false', () => {
    const wrapper = mount(<Comp fetch={{
      loading: false,
      error: null,
    }}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('will get data if needed', () => {
    const fetchQuestions = jest.fn();
    mount(<Comp
      fetch={{
        loading: false,
        error: null,
      }}
      shouldFetch
      fetchQuestions={fetchQuestions}
    />, {
      lifecycleExperimental: true,
    });
    expect(fetchQuestions).toHaveBeenCalled();
  });
});
