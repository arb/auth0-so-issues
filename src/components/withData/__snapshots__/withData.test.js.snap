// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<withData /> renders the component only if there are no errors and loading is false 1`] = `
<_class
  fetch={
    Object {
      "error": null,
      "loading": false,
    }
  }
>
  <Container
    name="test"
  >
    <div>
      <h1>
        test
      </h1>
      <Component
        fetch={
          Object {
            "error": null,
            "loading": false,
          }
        }
      >
        <div>
          hello world
        </div>
      </Component>
    </div>
  </Container>
</_class>
`;

exports[`<withData /> shows a spinner if loading is true 1`] = `
<Container
  name="test"
>
  <Spinner>
    <span>
      Loading data... Please be patient as there is a lot of data to load!
    </span>
  </Spinner>
</Container>
`;

exports[`<withData /> shows an error if present 1`] = `
<Container
  name="test"
>
  <span>
    There was an error getting data
  </span>
</Container>
`;
